### ISSUE TITLE
Openstack methods assume the extension for floating IP's is loaded

### ISSUE THREAD

**Issue by mikelococo** [fog/fog#2160]
server.public_ip_address and server.private_ip_address fail when executed against an Openstack implementation without the floating-ip extension loaded. I'm not sure how common this is, but Rackspace does it. While fog has a separate Rackspace provider, other small openstack implementations may follow their lead on config options.

Stacktraces:

    irb(main):006:0> server.public_ip_address
    Fog::Compute::OpenStack::NotFound: Fog::Compute::OpenStack::NotFound
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/excon-0.25.3/lib/excon/middlewares/expects.rb:10:in `response_call'
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/excon-0.25.3/lib/excon/middlewares/response_parser.rb:8:in `response_call'
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/excon-0.25.3/lib/excon/connection.rb:349:in `response'
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/excon-0.25.3/lib/excon/connection.rb:247:in `request'
    from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/fog-1.15.0/lib/fog/core/connection.rb:57:in `request'
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/fog-1.15.0/lib/fog/core/deprecated/connection.rb:20:in `request'
       from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/fog-1.15.0/lib/fog/openstack/compute.rb:339:in `request'
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/fog-1.15.0/lib/fog/openstack/requests/compute/list_all_addresses.rb:10:in `list_all_addresses'
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/fog-1.15.0/lib/fog/openstack/models/compute/server.rb:95:in `all_addresses'
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/fog-1.15.0/lib/fog/openstack/models/compute/server.rb:110:in `floating_ip_addresses'
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/fog-1.15.0/lib/fog/openstack/models/compute/server.rb:116:in `floating_ip_address'
        from (irb):6
        from /usr/local/bin/irb:12:in `<main>'

    irb(main):007:0> server.private_ip_address
    Fog::Compute::OpenStack::NotFound: Fog::Compute::OpenStack::NotFound
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/excon-0.25.3/lib/excon/middlewares/expects.rb:10:in `response_call'
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/excon-0.25.3/lib/excon/middlewares/response_parser.rb:8:in `response_call'
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/excon-0.25.3/lib/excon/connection.rb:349:in `response'
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/excon-0.25.3/lib/excon/connection.rb:247:in `request'
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/fog-1.15.0/lib/fog/core/connection.rb:57:in `request'
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/fog-1.15.0/lib/fog/core/deprecated/connection.rb:20:in `request'
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/fog-1.15.0/lib/fog/openstack/compute.rb:339:in `request'
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/fog-1.15.0/lib/fog/openstack/requests/compute/list_all_addresses.rb:10:in `list_all_addresses'
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/fog-1.15.0/lib/fog/openstack/models/compute/server.rb:95:in `all_addresses'
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/fog-1.15.0/lib/fog/openstack/models/compute/server.rb:110:in `floating_ip_addresses'
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/fog-1.15.0/lib/fog/openstack/models/compute/server.rb:122:in `private_ip_addresses'
        from /usr/local/Cellar/ruby/1.9.3-p392/lib/ruby/gems/1.9.1/gems/fog-1.15.0/lib/fog/openstack/models/compute/server.rb:126:in `private_ip_address'
        from (irb):7
        from /usr/local/bin/irb:12:in `<main>'

This was discovered while testing kitchen-openstack: https://github.com/RoboticCheese/kitchen-openstack/issues/16

**Comment by gshively11** [comment]
We're blocked by this issue.  chef-client blows up with an Excon::Errors:Forbidden error, due to this line: https://github.com/fog/fog/blob/639c5e348a5ea8757c8ea1cae1af151d063130c7/lib/fog/openstack/models/compute/server.rb#L147
Explicitly telling fog to use the private ip for ssh still attempts to grab the list of floating ips, which is not supported in our installation of OpenStack.  Any suggestions?

**Comment by krames** [comment]
@gshively11 As you probably know there are different strategies for allocating ip address in OpenStack and at the time this issue first occurred there was not a good way to programmatically retrieve an accessible ip address. I believe OpenStack was aware of this issue and was trying to address it, however, I haven't kept up with it.

TLDR; I do not think fog/chef is going to work out of the box for you. You might need to submit a pr to fix this.

@ @mdarby is this something you can help with?

**Comment by wolf31o2** [comment]
We have similar issues with caskdata/coopr-provisioner and the fog_provider plugin.

**Comment by fifieldt** [comment]
Coming up in OpenStack, the "don't care what you have to do, just get me a network" API:

http://specs.openstack.org/openstack/neutron-specs/specs/liberty/get-me-a-network.html
https://blueprints.launchpad.net/neutron/+spec/get-me-a-network

eta April 2016 :'(

**Comment by plribeiro3000** [comment]
Closing in favor of fog/fog-openstack#38
